File: color_sort_controller.c
//Header Files
#include <webots/distance_sensor.h>
#include <webots/motor.h>
#include <webots/position_sensor.h>
#include <webots/robot.h>
#include <webots/camera.h>
#include <webots/camera_recognition_object.h>

#include <stdio.h>

#define TIME_STEP 32

// 5 Cases for switch statement
enum State { WAITING, GRASPING, ROTATING, RELEASING, ROTATING_BACK };

int main(int argc, char **argv) {
  wb_robot_init();
  int counter = 0, i = 0;
  int state = WAITING;

  // crate target joint positions (same as before)
  const double target_positions[]  = {-1.88, -2.14, -2.38, -1.51};
  const double target_positions1[] = {-1.88,  2.14, -2.38, -1.51};
  const double target_positions2[] = {-1.88, -1.14, -2.38, -1.51};

  // joint angles when the gripper is over the box on the belt

  const double pickup_positions[]  = {-1.88, -1.60, -2.38, -1.51};

  double speed = 1.0;

  if (argc == 2)
    sscanf(argv[1], "%lf", &speed);

  // Gripper internal motors
  WbDeviceTag hand_motors[3];
  hand_motors[0] = wb_robot_get_device("finger_1_joint_1");
  hand_motors[1] = wb_robot_get_device("finger_2_joint_1");
  hand_motors[2] = wb_robot_get_device("finger_middle_joint_1");

  // ur5e joints
  WbDeviceTag ur_motors[4];
  ur_motors[0] = wb_robot_get_device("shoulder_lift_joint");
  ur_motors[1] = wb_robot_get_device("elbow_joint");
  ur_motors[2] = wb_robot_get_device("wrist_1_joint");
  ur_motors[3] = wb_robot_get_device("wrist_2_joint");

  // Camera
  WbDeviceTag camera = wb_robot_get_device("camera");
  wb_camera_enable(camera, 2 * TIME_STEP);
  wb_camera_recognition_enable(camera, 2 * TIME_STEP);

  double a = 0.0, b = 0.0, c = 0.0; // For storing the recognized RGB values from camera

  // Initial rotational motor speed
  for (i = 0; i < 4; ++i)
    wb_motor_set_velocity(ur_motors[i], speed);

  // Distance sensor
  WbDeviceTag distance_sensor = wb_robot_get_device("distance sensor");
  wb_distance_sensor_enable(distance_sensor, TIME_STEP);

  // Position sensor
  WbDeviceTag position_sensor = wb_robot_get_device("wrist_1_joint_sensor");
  wb_position_sensor_enable(position_sensor, TIME_STEP);

  while (wb_robot_step(TIME_STEP) != -1) {
    int number_of_objects = wb_camera_recognition_get_number_of_objects(camera);
    printf("\nRecognized %d objects.\n", number_of_objects);

    // Get and display all the objects information especially the color
    const WbCameraRecognitionObject *objects = wb_camera_recognition_get_objects(camera);
    for (i = 0; i < number_of_objects; ++i) {
      for (int j = 0; j < objects[i].number_of_colors; ++j) {
        a = objects[i].colors[3 * j];
        b = objects[i].colors[3 * j + 1];
        c = objects[i].colors[3 * j + 2];
        printf("%lf %lf %lf", a, b, c);
        printf("Color : %lf %lf %lf\n",
               objects[i].colors[3 * j],
               objects[i].colors[3 * j + 1],
               objects[i].colors[3 * j + 2]);
      }
    }

    // Switch cases for the different states of the arm
    if (counter <= 0) {
      switch (state) {
        case WAITING: {
          double d = wb_distance_sensor_get_value(distance_sensor);
          printf("WAITING, distance sensor = %lf\n", d);

          // condition: block close enough in front of the robot
          if (1) {//d < 500
            printf("Block detected → moving arm to pickup pose\n");
            // move arm so the gripper is over the box
            for (i = 0; i < 4; ++i)
              wb_motor_set_position(ur_motors[i], pickup_positions[i]);
            state = GRASPING;
            counter = 8;  // give time to reach pickup pose
          }
          break;
        }

        case GRASPING:
          // Now we assume arm is at pickup pose → close fingers
          printf("Grasping object (closing fingers)\n");
          for (i = 0; i < 3; ++i)
            wb_motor_set_position(hand_motors[i], 0.85);

          // Decide which crate to go to based on color
          if (a == 0.666667 && b == 1.000000 && c == 0.000000) {
            for (i = 0; i < 4; ++i)
              wb_motor_set_position(ur_motors[i], target_positions1[i]);
          } else if (a == 1.000000 && b == 0.000000 && c == 0.000000) {
            for (i = 0; i < 4; ++i)
              wb_motor_set_position(ur_motors[i], target_positions2[i]);
          } else {
            for (i = 0; i < 4; ++i)
              wb_motor_set_position(ur_motors[i], target_positions[i]);
          }

          printf("Rotating arm\n");
          state = ROTATING;
          break;

        case ROTATING:
          if (wb_position_sensor_get_value(position_sensor) < -2.3) {
            counter = 8;
            printf("Releasing object\n");
            state = RELEASING;
            for (i = 0; i < 3; ++i)
              wb_motor_set_position(hand_motors[i], wb_motor_get_min_position(hand_motors[i]));
          }
          break;

        case RELEASING:
          for (i = 0; i < 4; ++i)
            wb_motor_set_position(ur_motors[i], 0.0);
          printf("Rotating arm back\n");
          state = ROTATING_BACK;
          break;

        case ROTATING_BACK:
          if (wb_position_sensor_get_value(position_sensor) > -0.1) {
            state = WAITING;
            printf("Waiting object\n");
          }
          break;
      }
    }
    counter--;
  }

  wb_robot_cleanup();
  return 0;
}

File: color_sort_controller.py
from controller import Robot
import math
import numpy as np

from arm_ik import ArmIK   # <-- use the numeric IK class

TIME_STEP = 32
robot = Robot()

arm_ik = ArmIK()

# ---- DEVICE LIST DEBUGGER ----
print("\n=== DEVICE LIST START ===")
for i in range(robot.getNumberOfDevices()):
    dev = robot.getDeviceByIndex(i)
    try:
        print(i, dev.getName(), type(dev).__name__)
    except:
        print(i, "<unnamed>", type(dev).__name__)
print("=== DEVICE LIST END ===\n")
# ---- END DEBUGGER ----

# ---- STATES ----
WAITING, ANALYZING, MOVING_DROP, RELEASING, MOVING_HOME = range(5)

counter = 0
state = WAITING

# ---- CARTESIAN DROP POSES (TUNE THESE NUMBERS) ----
# These are *guesses* for world-frame TCP poses above each crate.
# You may need to tweak x,y,z slightly so the wrist ends up above each box.
GREEN_DROP_POSE = {
    "position": [0.55,  0.00, 0.30],      # center crate
    "rpy":      [0.0, math.pi, 0.0],      # wrist mostly down
}

RED_DROP_POSE = {
    "position": [0.55, -0.25, 0.30],      # right crate (neg Y)
    "rpy":      [0.0, math.pi, 0.0],
}

BLUE_DROP_POSE = {
    "position": [0.55,  0.25, 0.30],      # left crate (pos Y)
    "rpy":      [0.0, math.pi, 0.0],
}

# how long to wait for motions (in controller steps)
MOVE_TO_DROP_STEPS = 60    # ~2 seconds
MOVE_HOME_STEPS    = 60
OPEN_CLOSE_STEPS   = 20

speed = 1.0

# ----- devices -----

# Gripper internal motors
hand_motors = [
    robot.getDevice("finger_1_joint_1"),
    robot.getDevice("finger_2_joint_1"),
    robot.getDevice("finger_middle_joint_1"),
]

for m in hand_motors:
    m.setVelocity(2.0)

# Use actual joint limits to avoid warnings
GRIPPER_OPEN  = hand_motors[0].getMinPosition()   # ≈ 0.0495
GRIPPER_CLOSE = hand_motors[0].getMaxPosition()   # ≈ 0.85

# UR5e joints (6 DoF)
arm_joint_names = [
    "shoulder_pan_joint",
    "shoulder_lift_joint",
    "elbow_joint",
    "wrist_1_joint",
    "wrist_2_joint",
    "wrist_3_joint",
]

arm_motors = []
arm_sensors = []

for name in arm_joint_names:
    m = robot.getDevice(name)
    arm_motors.append(m)
    m.setVelocity(speed)
    ps = m.getPositionSensor()
    ps.enable(TIME_STEP)
    arm_sensors.append(ps)

# Camera
camera = robot.getDevice("camera")
camera.enable(2 * TIME_STEP)

# Distance sensor
distance_sensor = robot.getDevice("distance sensor")
distance_sensor.enable(TIME_STEP)

# Position sensor on wrist_1 (for "home" detection; home ≈ 0)
position_sensor = robot.getDevice("wrist_1_joint_sensor")
position_sensor.enable(TIME_STEP)

# Snapshot the initial joint config as "home" for going back later
# (We do this after one step, but initial is usually fine.)
HOME_Q = None

# Color reference values
RED_COLOR =   {'r': 0.904295, 'g': 0.243366, 'b': 0.14963}
BLUE_COLOR =  {'r': 0.037842, 'g': 0.03833,  'b': 0.904295}
GREEN_COLOR = {'r': 0.398413, 'g': 0.85156,  'b': 0.270146}

detected_color = None
last_block_present = False      # edge detect on distance sensor


# ---------- COLOR DETECTION ----------

def get_dominant_color(camera):
    """Analyze camera image and return 'red' / 'blue' / 'green'."""
    img = camera.getImage()
    if img is None:
        return None

    width = camera.getWidth()
    height = camera.getHeight()

    r_sum = g_sum = b_sum = 0.0
    samples = 0

    for y in range(height // 3, 2 * height // 3):
        for x in range(width // 3, 2 * width // 3):
            r = camera.imageGetRed(img, width, x, y) / 255.0
            g = camera.imageGetGreen(img, width, x, y) / 255.0
            b = camera.imageGetBlue(img, width, x, y) / 255.0
            r_sum += r
            g_sum += g
            b_sum += b
            samples += 1

    if samples == 0:
        return None

    avg_r = r_sum / samples
    avg_g = g_sum / samples
    avg_b = b_sum / samples

    def color_distance(c1, c2):
        return math.sqrt(
            (c1['r'] - c2['r'])**2 +
            (c1['g'] - c2['g'])**2 +
            (c1['b'] - c2['b'])**2
        )

    current = {'r': avg_r, 'g': avg_g, 'b': avg_b}

    red_dist   = color_distance(current, RED_COLOR)
    blue_dist  = color_distance(current, BLUE_COLOR)
    green_dist = color_distance(current, GREEN_COLOR)

    m = min(red_dist, blue_dist, green_dist)
    if m == red_dist:
        return 'red'
    elif m == blue_dist:
        return 'blue'
    else:
        return 'green'


# ---------- IK HELPERS ----------

def rpy_to_matrix(roll, pitch, yaw):
    cr, sr = math.cos(roll), math.sin(roll)
    cp, sp = math.cos(pitch), math.sin(pitch)
    cy, sy = math.cos(yaw), math.sin(yaw)

    # R = Rz(yaw) * Ry(pitch) * Rx(roll)
    R = np.array([
        [cy*cp, cy*sp*sr - sy*cr, cy*sp*cr + sy*sr],
        [sy*cp, sy*sp*sr + cy*cr, sy*sp*cr - cy*sr],
        [  -sp,            cp*sr,            cp*cr]
    ])
    return R


def make_pose_matrix(position, rpy):
    x, y, z = position
    roll, pitch, yaw = rpy
    R = rpy_to_matrix(roll, pitch, yaw)
    T = np.eye(4)
    T[0:3, 0:3] = R
    T[0:3, 3]   = [x, y, z]
    return T


def goto_pose_with_ik(position, rpy):
    """Compute IK for a pose and send all 6 joint commands."""
    T = make_pose_matrix(position, rpy)

    # read current q from sensors
    q_current = [ps.getValue() for ps in arm_sensors]

    # run IK with initial guess
    q_sol = arm_ik.computeInverseKinematics(T, q_current)

    # send to motors
    for i, m in enumerate(arm_motors):
        m.setPosition(float(q_sol[i]))


# ----- MAIN LOOP -----
first_step = True

while robot.step(TIME_STEP) != -1:

    # On first iteration, store HOME_Q as whatever pose the arm starts in
    if first_step:
        HOME_Q = [ps.getValue() for ps in arm_sensors]
        first_step = False

    if counter <= 0:
        if state == WAITING:
            d = distance_sensor.getValue()
            wrist_angle = position_sensor.getValue()
            print(f"WAITING, d = {d}, wrist = {wrist_angle}")

            # Keep gripper open while idle
            for m in hand_motors:
                m.setPosition(GRIPPER_OPEN)

            block_present = (d < 500)
            at_home = abs(wrist_angle) < 0.1

            # Only trigger when a NEW block arrives while arm is at home
            if block_present and at_home and not last_block_present:
                print("NEW block under gripper → CLOSE + ANALYZE")

                # Close immediately while block is still under gripper
                for m in hand_motors:
                    m.setPosition(GRIPPER_CLOSE)

                state = ANALYZING
                counter = OPEN_CLOSE_STEPS  # give time for fingers to close

            last_block_present = block_present

        elif state == ANALYZING:
            color = get_dominant_color(camera)
            if color is None:
                print("Color detection failed, default to GREEN")
                detected_color = 'green'
            else:
                detected_color = color
            print(f"Color detected: {detected_color}")

            # choose drop pose based on color
            if detected_color == 'red':
                pose = RED_DROP_POSE
                print("Moving to RED crate (IK)")
            elif detected_color == 'blue':
                pose = BLUE_DROP_POSE
                print("Moving to BLUE crate (IK)")
            else:
                pose = GREEN_DROP_POSE
                print("Moving to GREEN crate (IK)")

            # IK → joint commands
            goto_pose_with_ik(pose["position"], pose["rpy"])

            state = MOVING_DROP
            counter = MOVE_TO_DROP_STEPS

        elif state == MOVING_DROP:
            # done waiting for move → release
            print("Releasing crate")
            for m in hand_motors:
                m.setPosition(GRIPPER_OPEN)

            state = RELEASING
            counter = OPEN_CLOSE_STEPS

        elif state == RELEASING:
            # send arm back to home joint configuration
            print("Rotating arm back to home")
            for i, m in enumerate(arm_motors):
                m.setPosition(HOME_Q[i])

            state = MOVING_HOME
            counter = MOVE_HOME_STEPS

        elif state == MOVING_HOME:
            # after going home, reset state
            print("Ready for next crate\n")
            state = WAITING
            detected_color = None
            counter = 0  # just in case

    counter -= 1
